### ISSUE TITLE
Therealssh hangs after chacha20poly1305 error

### ISSUE THREAD

**Issue by ayuryshev** [skycoin/skywire-testnet#355]
# How to reproduce bug

1. Start 3 skywire-nodes: 
- NodeA with thereallssh app
- intermediary NodeB without apps
- NodeC with therealssh-client app
2. Add transports from NodeA and NodeC to NodeB
3. Start ssh communication. E.g. `therealssh-cli $PK_A "echo A"` (or `cat` a file, `ls`, `tree`, etc)
4. Increase amount of data. On some point of time NodeC logs:

```
[2019-05-16T12:19:03+03:00] WARN [router]: Stopped serving Transport: chacha20poly1305: message authentication failed
[2019-05-16T12:19:03+03:00] INFO [trmanager]: Transport 38254102-9309-5c24-85c0-d3098eb81256 failed with error: chacha20poly1305: message authentication failed. Re-dialing...
[2019-05-16T12:19:03+03:00] INFO [trmanager]: De-registered transport 38254102-9309-5c24-85c0-d3098eb81256
```

5. Stop NodeC. On exit NodeC logs:
```

[2019-05-16T12:24:21+03:00] INFO [router]: App request with type CreateLoop failed: route setup: EOF
[2019-05-16T12:24:21+03:00] INFO [trmanager]: Transport f422406e-2631-5659-98fe-1c277574063b failed with error: EOF. Re-dialing...
[2019-05-16T12:24:21+03:00] INFO [trmanager]: Closing transport manager
[2019-05-16T12:24:21+03:00] INFO [trmanager]: Transport f422406e-2631-5659-98fe-1c277574063b closed
[2019-05-16T12:24:21+03:00] INFO [trmanager]: Transport f422406e-2631-5659-98fe-1c277574063b closed
.....
[2019-05-16T12:24:21+03:00] INFO [trmanager]: Transport f422406e-2631-5659-98fe-1c277574063b closed
[2019-05-16T12:24:21+03:00] INFO [trmanager]: Transport f422406e-2631-5659-98fe-1c277574063b closed
[2019-05-16T12:24:21+03:00] INFO [trmanager]: Transport f422406e-2631-5659-98fe-1c277574063b closed
[2019-05-16T12:24:21+03:00] INFO [messenger]: Opened new link with the server 035915c609f71d0c7df27df85ec698ceca0cb262590a54f732e3bbd0cc68d89282
[2019-05-16T12:24:21+03:00] INFO [messenger]: Updating discovery entry
[2019-05-16T12:24:22+03:00] INFO [messenger]: Opened new channel local ID 0, remote ID 0 with 02603d53d49b6575a0b8cee05b70dd23c86e42cd6cb
```

5. Kill NodeB. It will log:
```
[2019-05-16T12:27:49+03:00] INFO [trmanager]: Closing transport manager
[2019-05-16T12:27:49+03:00] INFO [trmanager]: Transport a13b7256-8c99-5a47-b82c-982d9edb5442 closed
[2019-05-16T12:27:49+03:00] INFO [trmanager]: Transport a13b7256-8c99-5a47-b82c-982d9edb5442 closed
.... repeated lines ommitted ...
[2019-05-16T12:27:49+03:00] INFO [trmanager]: Transport a13b7256-8c99-5a47-b82c-982d9edb5442 closed
``` 
6. Kill NodeA. Note repeating log lines for closing the same transport

**Comment by ivcosla** [comment]
While working on my PR on this issue I got the following data race, referencing regions of the code that I have not changed, so it might also be something to look after in mainnet:

```bash
==================                                                                                                                 
WARNING: DATA RACE                                                                                                                 
Write at 0x00c0001bf9e0 by goroutine 75:                                                                                           
  bytes.(*Buffer).grow()                                                                                                           
      /usr/lib/go/src/bytes/buffer.go:101 +0x5a5                                                                                   
  bytes.(*Buffer).Write()                                                                                                          
      /usr/lib/go/src/bytes/buffer.go:175 +0x155                                                                                   
  github.com/skycoin/skywire/pkg/messaging.(*channel).readEncrypted()                                                              
      /home/ivan/skycoin/skywire/pkg/messaging/channel.go:244 +0x3bd                                                               
  github.com/skycoin/skywire/pkg/messaging.(*channel).Read()                                                                       
      /home/ivan/skycoin/skywire/pkg/messaging/channel.go:117 +0x3f1                                                               
  github.com/skycoin/skywire/pkg/transport.(*ManagedTransport).Read()                                                              
      /home/ivan/skycoin/skywire/pkg/transport/managed_transport.go:42 +0xbd                                                       
  io.ReadAtLeast()                                                                                                                 
      /usr/lib/go/src/io/io.go:310 +0x95                                                                                           
  github.com/skycoin/skywire/pkg/setup.(*Protocol).ReadPacket()                                                                    
      /usr/lib/go/src/io/io.go:329 +0xeb                                                                                           
  github.com/skycoin/skywire/pkg/setup.readAndDecodePacket()                                                                       
      /home/ivan/skycoin/skywire/pkg/setup/protocol.go:190 +0x3c                                                                   
  github.com/skycoin/skywire/pkg/setup.CreateLoop()                                                                                
      /home/ivan/skycoin/skywire/pkg/setup/protocol.go:149 +0x7c                                                                   
  github.com/skycoin/skywire/pkg/router.(*Router).requestLoop()                                                                    
      /home/ivan/skycoin/skywire/pkg/router/router.go:332 +0x123a                                                                  
  github.com/skycoin/skywire/pkg/router.(*Router).requestLoop-fm()                                                                 
      /home/ivan/skycoin/skywire/pkg/router/router.go:285 +0x59                                                                    
  github.com/skycoin/skywire/pkg/router.(*appManager).setupLoop()                                                                  
      /home/ivan/skycoin/skywire/pkg/router/app_manager.go:80 +0x17d                                                               
  github.com/skycoin/skywire/pkg/router.(*appManager).Serve.func1()                                                                
      /home/ivan/skycoin/skywire/pkg/router/app_manager.go:35 +0x3ed                                                               
  github.com/skycoin/skywire/pkg/app.(*Protocol).Serve.func1()                                                                     
      /home/ivan/skycoin/skywire/pkg/app/protocol.go:116 +0x90                                                                     
                                                                                                                                   
Previous read at 0x00c0001bf9e0 by goroutine 86:                                                                                   
  github.com/skycoin/skywire/pkg/messaging.(*channel).Read()                                                                       
      /usr/lib/go/src/bytes/buffer.go:76 +0x92                                                                                     
  github.com/skycoin/skywire/pkg/transport.(*ManagedTransport).Read()                                                              
      /home/ivan/skycoin/skywire/pkg/transport/managed_transport.go:42 +0xbd                                                       
  io.ReadAtLeast()                                                                                                                 
      /usr/lib/go/src/io/io.go:310 +0x95                                                                                           
  github.com/skycoin/skywire/pkg/setup.(*Protocol).ReadPacket()                                                                    
      /usr/lib/go/src/io/io.go:329 +0xeb                                                                                           
  github.com/skycoin/skywire/pkg/router.(*routeManager).Serve()                                                                    
      /home/ivan/skycoin/skywire/pkg/router/route_manager.go:78 +0x7c                                                              
  github.com/skycoin/skywire/pkg/router.(*Router).Serve.func1.1()                                                                  
      /home/ivan/skycoin/skywire/pkg/router/router.go:88 +0xc9                                                                     
                                                                                                                                   
Goroutine 75 (running) created at:                                                                                                 
  github.com/skycoin/skywire/pkg/app.(*Protocol).Serve()                                                                           
      /home/ivan/skycoin/skywire/pkg/app/protocol.go:110 +0x1cd                                                                    
  github.com/skycoin/skywire/pkg/router.(*appManager).Serve()                                                                      
      /home/ivan/skycoin/skywire/pkg/router/app_manager.go:29 +0xb4                                                                
  github.com/skycoin/skywire/pkg/router.(*Router).ServeApp()                                                                       
      /home/ivan/skycoin/skywire/pkg/router/router.go:155 +0x62d                                                                   
  github.com/skycoin/skywire/pkg/node.(*Node).SpawnApp.func2()                                                                     
      /home/ivan/skycoin/skywire/pkg/node/node.go:369 +0x1d0                                                                       
                                                                                                                                   
Goroutine 86 (running) created at:                                                                                                 
  github.com/skycoin/skywire/pkg/router.(*Router).Serve.func1()                                                                    
      /home/ivan/skycoin/skywire/pkg/router/router.go:84 +0x7a                                                                     
==================                                                                                                                 
==================                                                                                                                 
WARNING: DATA RACE
Write at 0x00c0001bf9f8 by goroutine 75:
  bytes.(*Buffer).grow()
      /usr/lib/go/src/bytes/buffer.go:102 +0x5c6
  bytes.(*Buffer).Write()
      /usr/lib/go/src/bytes/buffer.go:175 +0x155
  github.com/skycoin/skywire/pkg/messaging.(*channel).readEncrypted()
      /home/ivan/skycoin/skywire/pkg/messaging/channel.go:244 +0x3bd
  github.com/skycoin/skywire/pkg/messaging.(*channel).Read()
      /home/ivan/skycoin/skywire/pkg/messaging/channel.go:117 +0x3f1
  github.com/skycoin/skywire/pkg/transport.(*ManagedTransport).Read()
      /home/ivan/skycoin/skywire/pkg/transport/managed_transport.go:42 +0xbd
  io.ReadAtLeast()
      /usr/lib/go/src/io/io.go:310 +0x95
  github.com/skycoin/skywire/pkg/setup.(*Protocol).ReadPacket()
      /usr/lib/go/src/io/io.go:329 +0xeb
  github.com/skycoin/skywire/pkg/setup.readAndDecodePacket()
      /home/ivan/skycoin/skywire/pkg/setup/protocol.go:190 +0x3c
  github.com/skycoin/skywire/pkg/setup.CreateLoop()
      /home/ivan/skycoin/skywire/pkg/setup/protocol.go:149 +0x7c
  github.com/skycoin/skywire/pkg/router.(*Router).requestLoop()
      /home/ivan/skycoin/skywire/pkg/router/router.go:332 +0x123a 
  github.com/skycoin/skywire/pkg/router.(*Router).requestLoop-fm()
      /home/ivan/skycoin/skywire/pkg/router/router.go:285 +0x59
  github.com/skycoin/skywire/pkg/router.(*appManager).setupLoop() 
      /home/ivan/skycoin/skywire/pkg/router/app_manager.go:80 +0x17d
  github.com/skycoin/skywire/pkg/router.(*appManager).Serve.func1()
      /home/ivan/skycoin/skywire/pkg/router/app_manager.go:35 +0x3ed
  github.com/skycoin/skywire/pkg/app.(*Protocol).Serve.func1()
      /home/ivan/skycoin/skywire/pkg/app/protocol.go:116 +0x90

Previous read at 0x00c0001bf9f8 by goroutine 86:
  github.com/skycoin/skywire/pkg/messaging.(*channel).Read()
      /usr/lib/go/src/bytes/buffer.go:76 +0xad
  github.com/skycoin/skywire/pkg/transport.(*ManagedTransport).Read()
      /home/ivan/skycoin/skywire/pkg/transport/managed_transport.go:42 +0xbd
  io.ReadAtLeast()
      /usr/lib/go/src/io/io.go:310 +0x95
  github.com/skycoin/skywire/pkg/setup.(*Protocol).ReadPacket()
      /usr/lib/go/src/io/io.go:329 +0xeb
  github.com/skycoin/skywire/pkg/router.(*routeManager).Serve()
      /home/ivan/skycoin/skywire/pkg/router/route_manager.go:78 +0x7c
  github.com/skycoin/skywire/pkg/router.(*Router).Serve.func1.1() 
      /home/ivan/skycoin/skywire/pkg/router/router.go:88 +0xc9

Goroutine 75 (running) created at:
  github.com/skycoin/skywire/pkg/app.(*Protocol).Serve()
      /home/ivan/skycoin/skywire/pkg/app/protocol.go:110 +0x1cd
  github.com/skycoin/skywire/pkg/router.(*appManager).Serve()
      /home/ivan/skycoin/skywire/pkg/router/app_manager.go:29 +0xb4
  github.com/skycoin/skywire/pkg/router.(*Router).ServeApp()
      /home/ivan/skycoin/skywire/pkg/router/router.go:155 +0x62d
  github.com/skycoin/skywire/pkg/node.(*Node).SpawnApp.func2()
      /home/ivan/skycoin/skywire/pkg/node/node.go:369 +0x1d0

Goroutine 86 (running) created at:
  github.com/skycoin/skywire/pkg/router.(*Router).Serve.func1()
      /home/ivan/skycoin/skywire/pkg/router/router.go:84 +0x7a
```

**Comment by jdknives** [comment]
Closed by  #379
